EIPS/eip-5283.md

```diff
@@ -0,0 +1,112 @@
+
+Several EVM-based blockchains have successfully tested designs for the parallelization of the EVM. The best results have been obtained with fine-grained parallelization where conflicts are detected by tracking writes and reads of individual storage cells. The designs based on tracking the use of accounts or contracts provide only minor benefits because most transactions use the same [EIP-20](./eip-20.md) contracts.
+
+To summarize, the only available RPG construction today is based on using a contract storage cell. This construction is clean but it is not forward-compatible with transaction execution parallelization.
+
+
+## Specification
+
+Starting from an activation block (TBD) a new precompiled contract `Semaphore` is created at address `0x0A`. When `Semaphore` is called, if the caller address is present more than once in the call stack, the contract behaves as if the first instruction had been a `REVERT`, therefore the CALL returns 0. Otherwise, it executes no code and returns 1. The gas cost of the contract execution is set to 100, which is consumed independently of the call result.
+
+
+## Rationale
+
+
+The address `0x0A` is the next one available within the range defined by [EIP-1352](./eip-1352).
+
+### Sample usage
+
+```
+pragma solidity ^0.8.0;
+
+abstract contract ReentrancyGuard2 {
+
+    uint8 constant SemaphoreAddress = 0x0A;
+    /**
+     * @dev Prevents a contract from calling itself, directly or indirectly.
+     * Calling a `nonReentrant` function from another `nonReentrant`
+     * function is supported.      
+     */
+    modifier nonReentrant() {
+        _nonReentrantBefore();
+        _;
+    }
+
+    function _nonReentrantBefore() private {
+    	assembly {
+            if iszero(staticcall(1000,SemaphoreAddress,  0, 0, 0, 0)) {
+                revert(0, 0)
+            }
+        }
+    }
+}
+```
+
+
+### Parallellizable storage-based RPGs
+ 
+The only way to paralellize prexistent contracts that are using the storage RPG construction is that the VM automatically detects that a storage variable is used for the RPG, and proves that is works as required. This requires static code analysys. This is difficult to implement in consensus for two reasons. First, the CPU cost of detection and/or proving may be high. Second, some contract functions may not be protected by the RPG, meaning that some execution paths do not alter the RPG, which may complicate proving. Therefore this proposal aims to protect future contracts and let them be parallelizable, rather than to paralellize already deployed ones.
+
+### Alternatives
+
+There are alternative designs to implement RPGs on the EVM:
+
+1.	Transient storage opcodes (`TLOAD`/`TSTORE`) provide contract state that is kept between calls in the same transaction but it is not committed to the world state afterward. These opcodes do not address the problem of fine-grained parallelization. 
```

Comment:
why doesn't this help with parallelization? transactions that use any amount of transient storage can be executed in parallel, even if they use the same slot

Reply:
is this about parallel execution for a single transaction?

Reply:
My mistake! I misplaced that sentence when rewritting the paragraph according to the last review.

Reply:
so i understand correctly, the transient storage eip provides the same benefits as this eip, and is more general, i.e. not limited to reentrancy locks? would be helpful to have a `discussions-to` url in the header early so there's a better place to discuss, but also think it should be explained in the alternate or rationale how this eip is different

Reply:
yes, transient opcodes are better from the EVM "purity" standpoint. But I haven't heard of any use for the transient opcodes apart from mutexes. I will add the discussions tag, where we should be chatting.